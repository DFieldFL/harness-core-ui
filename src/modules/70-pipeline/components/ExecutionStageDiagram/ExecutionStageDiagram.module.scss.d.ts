/* eslint-disable */
// this is an auto-generated file
declare const styles: {
  readonly main: string
  readonly whiteBackground: string
  readonly runningNode: string
  readonly shine: string
  readonly selected: string
  readonly stageSelection: string
  readonly canvas: string
  readonly groupLabels: string
  readonly label: string
  readonly selectedLabel: string
  readonly fadeIn: string
  readonly stageItem: string
  readonly stageButton: string
  readonly stageError: string
  readonly errorTooltip: string
  readonly rotate: string
}
export default styles
