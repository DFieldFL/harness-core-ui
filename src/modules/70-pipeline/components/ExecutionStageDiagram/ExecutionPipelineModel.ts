import type { CSSProperties } from 'react'
import type { IconName, SelectOption } from '@wings-software/uicore'
import type { ExecutionStatus } from '@pipeline/utils/statusHelpers'
import type { NodeRunInfo } from 'services/pipeline-ng'

// TODO: should be replaced with type-shape of the box (like: Stage, Step, Approval)
export enum ExecutionPipelineNodeType {
  DIAMOND = 'DIAMOND',
  NORMAL = 'NORMAL',
  ICON = 'ICON'
}

export interface StageOptions extends SelectOption {
  value: string
  disabled?: boolean
}

export interface ExecutionPipelineItem<T> {
  [x: string]: any
  iconStyle?: CSSProperties
  iconSize?: number
  identifier: string
  name: string
  type: ExecutionPipelineNodeType
  status: ExecutionStatus
  icon: IconName
  skipCondition?: string
  when?: NodeRunInfo
  barrierFound?: boolean
  /**
   * @default false
   */
  showInLabel?: boolean
  disableClick?: boolean
  cssProps?: CSSProperties
  data?: T
  pipeline?: ExecutionPipeline<T>
  itemType?: 'step' | 'service-dependency' | string
}

export interface ExecutionPipelineGroupInfo<T> {
  identifier: string
  name: string
  data: T
  cssProps?: CSSProperties
  icon: IconName
  skipCondition?: string
  when?: NodeRunInfo
  containerCss?: CSSProperties
  textCss?: CSSProperties
  verticalStepGroup?: boolean
  showInLabel?: boolean // Default = true
  status: ExecutionStatus
  isOpen: boolean
  items: Array<ExecutionPipelineNode<T>>
}
export interface ExecutionPipelineNode<T> {
  item?: ExecutionPipelineItem<T>
  parallel?: Array<ExecutionPipelineNode<T>>
  group?: ExecutionPipelineGroupInfo<T>
}

export interface ExecutionPipeline<T> {
  items: Array<ExecutionPipelineNode<T>>
  identifier: string
  status?: ExecutionStatus
  allNodes: string[]
}
