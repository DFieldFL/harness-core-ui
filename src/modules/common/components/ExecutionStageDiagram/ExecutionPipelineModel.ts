import type { IconName, SelectOption } from '@wings-software/uikit'

// TODO: should be replaced with type-shape of the box (like: Stage, Step, Approval)
export enum ExecutionPipelineNodeType {
  DIAMOND = 'DIAMOND',
  NORMAL = 'NORMAL'
}

// TODO: should be replaced or same as DTO
export enum ExecutionPipelineItemStatus {
  SUCCEEDED = 'Succeeded',
  FAILED = 'Failed',
  SUCCESS = 'Success',
  ABORTED = 'Aborted',
  ERROR = 'Error',
  PAUSED = 'Paused',
  PAUSING = 'Pausing',
  WAITING = 'Waiting',
  ABORTING = 'Aborting',
  RUNNING = 'Running',
  QUEUED = 'Queued',
  ROLLBACK = 'Rollback',
  SKIPPED = 'Skipped',
  STARTING = 'Starting',
  REJECTED = 'Rejected',
  EXPIRED = 'Expired',
  NOT_STARTED = 'NotStarted',
  ASYNC_WAITING = 'AsyncWaiting'
}

export interface StageOptions extends SelectOption {
  value: string
  disabled?: boolean
}

export interface ExecutionPipelineItem<T> {
  identifier: string
  name: string
  type: ExecutionPipelineNodeType
  status: ExecutionPipelineItemStatus
  icon: IconName
  showInLabel?: boolean // Default = false
  cssProps?: React.CSSProperties
  data?: T
  pipeline?: ExecutionPipeline<T>
}

export interface ExecutionPipelineGroupInfo<T> {
  identifier: string
  name: string
  data: T
  cssProps?: React.CSSProperties
  icon: IconName
  status: ExecutionPipelineItemStatus
  isOpen: boolean
  items: Array<ExecutionPipelineNode<T>>
}
export interface ExecutionPipelineNode<T> {
  item?: ExecutionPipelineItem<T>
  parallel?: Array<ExecutionPipelineNode<T>>
  group?: ExecutionPipelineGroupInfo<T>
}

export interface ExecutionPipeline<T> {
  items: Array<ExecutionPipelineNode<T>>
  identifier: string
  status?: ExecutionPipelineItemStatus
}
