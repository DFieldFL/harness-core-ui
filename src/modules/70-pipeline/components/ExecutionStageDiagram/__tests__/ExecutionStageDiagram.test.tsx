import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react'
import type { ExecutionStageDiagramProps } from '@pipeline/exports'
import ResizeObserver from './ResizeObserver'
import { ExecutionPipelineItemStatus, ExecutionPipelineNodeType } from '../ExecutionPipelineModel'
import ExecutionStageDiagram from '../ExecutionStageDiagram'

interface Data {
  label: string
}

const itemClickHandler = jest.fn()
const itemMouseEnter = jest.fn()
const itemMouseLeave = jest.fn()
const canvasListener = jest.fn()
const onChangeStageSelection = jest.fn()

// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
const getExtraProps = () => ({
  itemClickHandler,
  itemMouseEnter,
  itemMouseLeave,
  canvasListener,
  onChangeStageSelection
})

const getProps = (): ExecutionStageDiagramProps<Data> => ({
  data: {
    items: [
      {
        item: {
          icon: 'pipeline-deploy',
          identifier: 'qaStage',
          name: 'qa stage',
          status: ExecutionPipelineItemStatus.FAILED,
          type: ExecutionPipelineNodeType.NORMAL,
          data: {
            label: 'qaStage'
          }
        }
      },
      {
        parallel: [
          {
            item: {
              icon: 'pipeline-deploy',
              identifier: 'parallel1',
              name: 'Parallel 1',
              status: ExecutionPipelineItemStatus.PAUSED,
              type: ExecutionPipelineNodeType.NORMAL,
              data: {
                label: 'Parallel 1'
              }
            }
          },
          {
            item: {
              icon: 'pipeline-deploy',
              identifier: 'parallel3',
              name: 'Parallel 2',
              status: ExecutionPipelineItemStatus.SUCCESS,
              type: ExecutionPipelineNodeType.NORMAL,
              data: {
                label: 'Parallel 2'
              }
            }
          }
        ]
      },
      {
        item: {
          icon: 'pipeline-deploy',
          identifier: 'stage2',
          name: 'stage 2',
          status: ExecutionPipelineItemStatus.RUNNING,
          type: ExecutionPipelineNodeType.NORMAL,
          data: {
            label: 'stage2'
          }
        }
      },
      {
        group: {
          icon: 'service' as any,
          identifier: 'Service',
          name: 'service',
          isOpen: true,
          status: ExecutionPipelineItemStatus.WAITING,
          data: {
            label: 'service'
          },
          items: [
            {
              item: {
                icon: 'badge' as any,
                identifier: 'badge',
                name: 'Badge',
                status: ExecutionPipelineItemStatus.ABORTED,
                type: ExecutionPipelineNodeType.NORMAL,
                data: {
                  label: 'badge'
                }
              }
            },
            {
              item: {
                icon: 'barcode' as any,
                identifier: 'barcode',
                name: 'barcode',
                status: ExecutionPipelineItemStatus.NOT_STARTED,
                type: ExecutionPipelineNodeType.NORMAL,
                data: {
                  label: 'barcode'
                }
              }
            }
          ]
        }
      }
    ],
    identifier: 'Test_Pipline',
    status: ExecutionPipelineItemStatus.ERROR
  },
  selectedIdentifier: 'qaStage',
  showStageSelection: true,
  selectedStage: { label: 'QA', value: 'qa', icon: { name: 'add' } },
  stageSelectionOptions: [
    { label: 'QA', value: 'qa', icon: { name: 'add' } },
    { label: 'Prod', value: 'prod', icon: { name: 'minus' } }
  ],
  gridStyle: { startX: 50, startY: 50 }
})
jest.mock('resize-observer-polyfill', () => {
  return ResizeObserver
})

describe('Test Execution StageDiagram', () => {
  beforeAll(() => {
    jest.spyOn(global.Math, 'random').mockReturnValue(0.12345)
  })
  afterAll(() => {
    jest.spyOn(global.Math, 'random').mockReset()
  })
  test('should render the default snapshot', () => {
    const { container } = render(<ExecutionStageDiagram {...getProps()} />)
    expect(container).toMatchSnapshot()
  })
})

describe('Test Execution StageDiagram - Action/Events', () => {
  // eslint-disable-next-line jest/no-disabled-tests
  test.skip('Test Mouse Events on Nodes and Canvas', () => {
    const { container } = render(
      <ExecutionStageDiagram
        {...getProps()}
        {...getExtraProps()}
        diagramContainerHeight={100}
        selectedIdentifier="parallel1"
      />
    )
    const node = container.querySelector('[data-nodeid="qaStage"] .defaultNode') as HTMLElement
    fireEvent.mouseEnter(node)
    expect(itemMouseEnter).toBeCalled()
    fireEvent.mouseLeave(node)
    expect(itemMouseLeave).toBeCalled()
    fireEvent.click(node)
    expect(itemClickHandler).toBeCalled()
    const canvasButton = container.querySelectorAll('.canvasButtons button')[0]
    fireEvent.click(canvasButton)
    expect(canvasListener).toBeCalledWith(3)
  })

  test('Test click Event on label', async () => {
    const { container } = render(<ExecutionStageDiagram {...getProps()} {...getExtraProps()} />)
    const label = container.querySelector('.groupLabels .label') as HTMLElement
    fireEvent.click(label)
    waitFor(() => container.querySelector('.groupLabels .selectedLabel'))
    expect(label?.classList.contains('selectedLabel')).toBeTruthy()
  })

  test('Test Stage Selection', async () => {
    const { getByText, getByPlaceholderText } = render(<ExecutionStageDiagram {...getProps()} {...getExtraProps()} />)
    const qaStage = getByText('QA')
    fireEvent.click(qaStage)
    await waitFor(() => getByPlaceholderText('Filter...'))
    const search = getByPlaceholderText('Filter...')
    fireEvent.change(search, { target: { value: 'Prod' } })
    const prodStage = getByText('Prod')
    fireEvent.click(prodStage)
    expect(onChangeStageSelection).toBeCalledWith({ label: 'Prod', value: 'prod', icon: { name: 'minus' } })
  })
})
