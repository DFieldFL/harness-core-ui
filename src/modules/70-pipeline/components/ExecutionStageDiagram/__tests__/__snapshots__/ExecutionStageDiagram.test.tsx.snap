// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Execution StageDiagram should render the default snapshot 1`] = `
<div>
  <div
    class="main"
  >
    <div
      class="canvas canvas css-18f0176"
    >
      <div
        class="css-12q0bj3"
        style="transform: translate(
				0px,
				0px)
			scale(
				1
			);"
      />
      <div
        class="css-12q0bj3"
        style="transform: translate(
				0px,
				0px)
			scale(
				1
			);"
      >
        <div
          style="left: 1190px; cursor: pointer; top: 8px; pointer-events: none; position: absolute; height: 160px;"
        >
          <div
            class="groupLayer"
            style="width: 480px; height: 160px;"
          />
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main header"
          style="top: 13px; left: 1200px; width: 460px; align-items: center;"
        >
          <button
            class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--iconOnly Button--without-shadow"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-minus StyledProps--main"
              icon="minus"
            >
              <svg
                data-icon="minus"
                height="8"
                viewBox="0 0 16 16"
                width="8"
              >
                <desc>
                  minus
                </desc>
                <path
                  d="M13 7H3c-.55 0-1 .45-1 1s.45 1 1 1h10c.55 0 1-.45 1-1s-.45-1-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main"
            style="--text-line-clamp: 1;"
          >
            Step Group HTTP
             
          </p>
        </div>
        <div
          class="node css-nlpftr"
          data-nodeid="11111111-1111-4111-9111-111111111111"
          style="top: 50px; left: 1670px;"
        >
          <div
            class="emptyNode"
          >
            <div>
              <div
                class="port portIn"
              >
                <div
                  class="port "
                  data-name="In"
                  data-nodeid="11111111-1111-4111-9111-111111111111"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="port portOut"
              >
                <div
                  class="port "
                  data-name="Out"
                  data-nodeid="11111111-1111-4111-9111-111111111111"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-12q0bj3"
        style="transform: translate(
				0px,
				0px)
			scale(
				1
			);"
      >
        <div
          class="node css-nlpftr"
          data-nodeid="start-new"
          style="top: 50px; left: 50px;"
        >
          <div
            class="defaultNode"
          >
            <div
              class="nodeStart"
            >
              <div>
                <span
                  class="bp3-icon bp3-icon-play StyledProps--main icon"
                  icon="play"
                >
                  <svg
                    data-icon="play"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      play
                    </desc>
                    <path
                      d="M12 8c0-.35-.19-.64-.46-.82l.01-.02-6-4-.01.02A.969.969 0 005 3c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1 .21 0 .39-.08.54-.18l.01.02 6-4-.01-.02c.27-.18.46-.47.46-.82z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <div>
                  <div
                    style="visibility: initial;"
                  >
                    <div
                      class="port portOut"
                    >
                      <div
                        class="port "
                        data-name="Out"
                        data-nodeid="start-new"
                      >
                        <button
                          class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                          type="button"
                        >
                          <span
                            class="bp3-icon bp3-icon-circle StyledProps--main"
                            icon="circle"
                          >
                            <svg
                              data-icon="circle"
                              height="8"
                              viewBox="0 0 16 16"
                              width="8"
                            >
                              <desc>
                                circle
                              </desc>
                              <path
                                d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                  <div
                    style="visibility: hidden;"
                  />
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="node css-nlpftr"
          data-nodeid="$node$-$node$parallel1-Start"
          style="top: 50px; left: 350px;"
        >
          <div
            class="emptyNode"
          >
            <div>
              <div
                class="port portIn"
              >
                <div
                  class="port "
                  data-name="In"
                  data-nodeid="$node$-$node$parallel1-Start"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="port portOut"
              >
                <div
                  class="port "
                  data-name="Out"
                  data-nodeid="$node$-$node$parallel1-Start"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="node css-nlpftr"
          data-nodeid="$node$parallel1$node$-End"
          style="top: 50px; left: 630px;"
        >
          <div
            class="emptyNode"
          >
            <div>
              <div
                class="port portIn"
              >
                <div
                  class="port "
                  data-name="In"
                  data-nodeid="$node$parallel1$node$-End"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="port portOut"
              >
                <div
                  class="port "
                  data-name="Out"
                  data-nodeid="$node$parallel1$node$-End"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="node css-nlpftr"
          data-nodeid="stop-node"
          style="top: 50px; left: 2030px;"
        >
          <div
            class="defaultNode"
          >
            <div
              class="nodeStart"
            >
              <div>
                <span
                  class="bp3-icon bp3-icon-stop StyledProps--main icon"
                  icon="stop"
                >
                  <svg
                    data-icon="stop"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      stop
                    </desc>
                    <path
                      d="M12 3H4c-.55 0-1 .45-1 1v8c0 .55.45 1 1 1h8c.55 0 1-.45 1-1V4c0-.55-.45-1-1-1z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <div>
                  <div
                    style="visibility: hidden;"
                  />
                  <div
                    style="visibility: initial;"
                  >
                    <div
                      class="port portIn"
                    >
                      <div
                        class="port "
                        data-name="In"
                        data-nodeid="stop-node"
                      >
                        <button
                          class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                          type="button"
                        >
                          <span
                            class="bp3-icon bp3-icon-circle StyledProps--main"
                            icon="circle"
                          >
                            <svg
                              data-icon="circle"
                              height="8"
                              viewBox="0 0 16 16"
                              width="8"
                            >
                              <desc>
                                circle
                              </desc>
                              <path
                                d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                                fill-rule="evenodd"
                              />
                            </svg>
                          </span>
                        </button>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="css-12q0bj3"
        style="transform: translate(
				0px,
				0px)
			scale(
				1
			);"
      >
        <div
          style="left: 910px; cursor: pointer; top: -34px; pointer-events: none; position: absolute; height: 216px;"
        >
          <div
            class="groupLayer"
            style="width: 1040px; height: 216px;"
          />
        </div>
        <div
          class="Layout--horizontal Layout--layout-spacing-xsmall StyledProps--main header"
          style="top: -29px; left: 920px; width: 1020px; align-items: center;"
        >
          <button
            class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main Button--iconOnly Button--without-shadow"
            type="button"
          >
            <span
              class="bp3-icon bp3-icon-minus StyledProps--main"
              icon="minus"
            >
              <svg
                data-icon="minus"
                height="8"
                viewBox="0 0 16 16"
                width="8"
              >
                <desc>
                  minus
                </desc>
                <path
                  d="M13 7H3c-.55 0-1 .45-1 1s.45 1 1 1h10c.55 0 1-.45 1-1s-.45-1-1-1z"
                  fill-rule="evenodd"
                />
              </svg>
            </span>
          </button>
          <p
            class="StyledProps--font Text--lineclamp StyledProps--main"
            style="--text-line-clamp: 1;"
          >
            service
             
          </p>
        </div>
        <div
          class="node css-nlpftr"
          data-nodeid="11111111-1111-4111-9111-111111111111"
          style="top: 50px; left: 1950px;"
        >
          <div
            class="emptyNode"
          >
            <div>
              <div
                class="port portIn"
              >
                <div
                  class="port "
                  data-name="In"
                  data-nodeid="11111111-1111-4111-9111-111111111111"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
            <div>
              <div
                class="port portOut"
              >
                <div
                  class="port "
                  data-name="Out"
                  data-nodeid="11111111-1111-4111-9111-111111111111"
                >
                  <button
                    class="bp3-button bp3-minimal Button--button StyledProps--font StyledProps--main portBtn StyledProps--intent-primary Button--with-current-color Button--without-shadow"
                    type="button"
                  >
                    <span
                      class="bp3-icon bp3-icon-circle StyledProps--main"
                      icon="circle"
                    >
                      <svg
                        data-icon="circle"
                        height="8"
                        viewBox="0 0 16 16"
                        width="8"
                      >
                        <desc>
                          circle
                        </desc>
                        <path
                          d="M8 0C3.6 0 0 3.6 0 8s3.6 8 8 8 8-3.6 8-8-3.6-8-8-8zm0 14c-3.3 0-6-2.7-6-6s2.7-6 6-6 6 2.7 6 6-2.7 6-6 6z"
                          fill-rule="evenodd"
                        />
                      </svg>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <svg
        class="css-12q0bj3"
        style="transform: translate(
				0px,
				0px)
			scale(
				1
			);"
      />
    </div>
    <div
      class="Layout--horizontal Layout--layout-spacing-xxlarge StyledProps--main stageSelection"
    >
      <div
        class="groupLabels"
      >
        <span
          class="bp3-popover-wrapper"
        >
          <span
            class="bp3-popover-target"
          >
            <div
              class=""
            >
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main stageButton Button--with-current-color"
                type="button"
              >
                <span
                  class="bp3-icon bp3-icon-add StyledProps--main StyledProps--padding-right-xsmall"
                  icon="add"
                >
                  <svg
                    data-icon="add"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      add
                    </desc>
                    <path
                      d="M10.99 6.99h-2v-2c0-.55-.45-1-1-1s-1 .45-1 1v2h-2c-.55 0-1 .45-1 1s.45 1 1 1h2v2c0 .55.45 1 1 1s1-.45 1-1v-2h2c.55 0 1-.45 1-1s-.45-1-1-1zm-3-7c-4.42 0-8 3.58-8 8s3.58 8 8 8 8-3.58 8-8-3.58-8-8-8zm0 14c-3.31 0-6-2.69-6-6s2.69-6 6-6 6 2.69 6 6-2.68 6-6 6z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
                <span
                  class="bp3-button-text"
                >
                  QA
                </span>
                <span
                  class="bp3-icon bp3-icon-caret-down StyledProps--main"
                  icon="caret-down"
                >
                  <svg
                    data-icon="caret-down"
                    height="16"
                    viewBox="0 0 16 16"
                    width="16"
                  >
                    <desc>
                      caret-down
                    </desc>
                    <path
                      d="M12 6.5c0-.28-.22-.5-.5-.5h-7a.495.495 0 00-.37.83l3.5 4c.09.1.22.17.37.17s.28-.07.37-.17l3.5-4c.08-.09.13-.2.13-.33z"
                      fill-rule="evenodd"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </span>
        </span>
        <div
          class="main"
        />
      </div>
      <div
        class="groupLabels"
      >
        <span
          class="label"
        >
          service
        </span>
      </div>
    </div>
    <span
      class="canvasButtons"
    >
      <div
        class="Layout--vertical Layout--layout-spacing-medium StyledProps--main"
        id="button-group"
      >
        <div
          class="StyledProps--font StyledProps--main ButtonGroup--btnGroup"
        >
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly"
                tabindex="0"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="canvas-position"
                >
                  <svg
                    height="16"
                    viewBox="0 0 12 12"
                    width="16"
                  >
                    <path
                      d="M8 0h4v3.333h-1.333v-2H8V0zM4 0v1.333H1.333v2H0V0h4zm4 12v-1.333h2.667v-2H12V12H8zm-4 0H0V8.667h1.333v2H4V12z"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
        <div
          class="StyledProps--font StyledProps--main ButtonGroup--btnGroup"
        >
          <span
            class="bp3-popover-wrapper"
          >
            <span
              class="bp3-popover-target"
            >
              <button
                class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly"
                tabindex="0"
                type="button"
              >
                <span
                  class="bp3-icon StyledProps--main"
                  data-icon="canvas-selector"
                >
                  <svg
                    height="16"
                    viewBox="0 0 14 14"
                    width="16"
                  >
                    <path
                      d="M0 8h1v1H0V8zm0 5h1v-1H0v1zm0-2h1v-1H0v1zm0-4h1V6H0v1zm0-6h1V0H0v1zm0 4h1V4H0v1zm0-2h1V2H0v1zm12 6h1V8h-1v1zm1-3h-1v1h1V6zm-2-6h-1v1h1V0zm2 0h-1v1h1V0zm0 2h-1v1h1V2zm0 2h-1v1h1V4zM4 1h1V0H4v1zm8 13h1v-1h-1v1zM8 1h1V0H8v1zm2 11v1h2v-1h-2zm3-2h-1v2h1v-2zm0 2v1h1v-1h-1zm-9 1h1v-1H4v1zm-2 0h1v-1H2v1zM2 1h1V0H2v1zm6 12h1v-1H8v1zM6 1h1V0H6v1zm0 12h1v-1H6v1z"
                    />
                  </svg>
                </span>
              </button>
            </span>
          </span>
        </div>
        <span
          class="verticalButtons"
        >
          <div
            class="StyledProps--font StyledProps--main ButtonGroup--btnGroup"
          >
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon StyledProps--main"
                    data-icon="zoom-in"
                  >
                    <svg
                      height="16"
                      viewBox="0 0 10 10"
                      width="16"
                    >
                      <path
                        d="M9.61 4.61H5.39V.39a.39.39 0 10-.78 0v4.22H.39a.39.39 0 100 .78h4.22v4.22a.39.39 0 10.78 0V5.39h4.22a.39.39 0 100-.78z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
            <span
              class="bp3-popover-wrapper"
            >
              <span
                class="bp3-popover-target"
              >
                <button
                  class="bp3-button Button--button StyledProps--font StyledProps--main Button--with-current-color Button--iconOnly"
                  tabindex="0"
                  type="button"
                >
                  <span
                    class="bp3-icon StyledProps--main"
                    data-icon="zoom-out"
                  >
                    <svg
                      height="16"
                      viewBox="0 0 10 2"
                      width="16"
                    >
                      <path
                        d="M9.61.61H.39a.39.39 0 100 .78h9.22a.39.39 0 100-.78z"
                      />
                    </svg>
                  </span>
                </button>
              </span>
            </span>
          </div>
        </span>
      </div>
    </span>
  </div>
</div>
`;
